### PR TITLE
fix(#286): parity-doc coverage test + T025 + parity-gate retry

### LINKED ISSUE #286: Parity test: every CommandSpec in run_pr_preflight.py must appear in LOCAL_CI_PARITY_INVARIANTS.md
## Current state (verified 2026-04-18 against `main` + `ci/cleanup-parity-and-doc-gaps`)

`scripts/run_pr_preflight.py` declares 38 unique `CommandSpec` names. 16 of them do not appear as a double-quoted string anywhere in `LOCAL_CI_PARITY_INVARIANTS.md`. 22 appear at least once (21 via an exact `CommandSpec "X"` cite, 1 additional via the compound `CommandSpecs "A" and "B"` form used by Row 17).

Verification method: AST walk of `run_pr_preflight.py` collecting every `CommandSpec(...)` call's first positional-argument string literal, cross-referenced against every double-quoted string in the doc.

## The 16 missing names

- Baseline staleness (FR-025)
- Coverage delta parity (Codecov project status)
- Extension lint
- Extension test config parity
- Extension test count validation
- Extension test lint
- Extension test type check
- No typing.Any in src/, tests/, scripts/ (QG-40)
- Partial-branch ratchet
- Python test count validation
- Rule-disable invariants (FR-014)
- Suppression baseline sync gate
- Suppression justifications
- Suppression main-baseline gate
- Suppression scope coverage (FR-026)
- Test floor contract validation

## Categorization

**Group A — existing Tier 2 row covers the gate by behavior; Prevention Proof column does not cite the CommandSpec name (9):**

- Row 11 (suppression-audit) ↔ Suppression baseline sync gate, Suppression main-baseline gate, Suppression justifications
- Row 11a (scope-coverage full-tree) ↔ Suppression scope coverage (FR-026)
- Row 11b (baseline-staleness) ↔ Baseline staleness (FR-025)
- Row 11c (rule-disable-invariants full-tree) ↔ Rule-disable invariants (FR-014)
- Row 26 (test count Python) ↔ Python test count validation
- Row 27 (test count Extension) ↔ Extension test count validation
- Row 32 (coverage-delta) ↔ Coverage delta parity (Codecov project status)

**Group B — Tier 1 row covers the gate; preflight re-runs it at Tier 2 as pre-push reinforcement (4):**

- Row 7 (extension-tests ESLint) ↔ Extension lint
- Row 7a (extension-tests test tsc) ↔ Extension test type check
- Row 7b (test config parity) ↔ Extension test config parity
- Row 7c (extension-tests test ESLint) ↔ Extension test lint

**Group C — no corresponding row anywhere in the doc (3):**

- No typing.Any in src/, tests/, scripts/ (QG-40) — runs `scripts/check_no_any_types.py`
- Partial-branch ratchet — runs `pnpm --dir extension run test:partial-branches` (delegates to `scripts/check_partial_branches.py`)
- Test floor contract validation — runs `scripts/check_test_floor_contract.py`

## Existing infrastructure the new test can reuse

`tests/unit/test_ci_parity_drift.py` already exposes `_load_preflight_module()`, `_normalized_preflight_commands()`, and AST helpers. Classes in that file (`TestCliReferenceGateParity`, `TestPartialBranchesParity`, `TestRatchetBumpGuardParity`, `TestFormatCheckParity`, and others) verify individual CommandSpecs exist in preflight. None assert the doc covers them.

## Proposed test

Add `tests/unit/test_parity_doc_coverage.py` that:

1. AST-walks `scripts/run_pr_preflight.py` and collects every `CommandSpec` call's first positional-argument string literal. Reuse `_load_preflight_module` / `_normalized_preflight_commands` where practical.
2. Reads `LOCAL_CI_PARITY_INVARIANTS.md` and extracts every double-quoted string using a regex that handles both `CommandSpec "X"` and compound `CommandSpecs "A" and "B"` forms.
3. Asserts the CommandSpec name set from step 1 is a subset of the quoted-string set from step 2.
4. On failure, prints the missing names with a suggested Prevention Proof cite template.
5. Includes adversarial negative tests that prove the checker catches (a) a fabricated new CommandSpec without a doc cite and (b) a removed cite for an existing CommandSpec. Required by the "never claim enforcement without proof" standing invariant.

Cross-platform deterministic: AST + pathlib + re; no filesystem ordering.

## Design decisions required before implementation

1. **Exact-match vs fuzzy.** Exact-match (CommandSpec name appears verbatim as a double-quoted string) is durable and catches typos. Fuzzy (substring or token-overlap) is more forgiving for doc authors but invites drift.
2. **Whether a Tier 1 row satisfies the test for preflight re-runs.** Accepting Tier 1 rows means no new Tier 2 rows for Group B (4 items). Refusing them means writing 4 minimal Tier 2 acknowledgement rows.
3. **Orphan detection.** The previous issue text said "extras in the doc are OK". An orphan check (doc cites a CommandSpec name that no longer exists in `run_pr_preflight.py`) catches typos and dead cites; whether to treat orphans as a warning or a failure is a separate call.
4. **Test floor contract validation placement.** Either add a dedicated Tier 2 row for `check_test_floor_contract.py`, or extend Rows 26/27 to cite it alongside `validate-test-results.py`.

## Acceptance

- `tests/unit/test_parity_doc_coverage.py` runs in the standard pytest gate (pre-push + CI + `test:ci`) with no new entry point.
- Adding a new `CommandSpec` to `run_pr_preflight.py` without a corresponding doc entry fails the test, locally and in CI.
- All 16 currently-missing CommandSpecs gain a representation per the design decisions above: cite added to existing Tier 2 row (Group A, up to 9), Tier 1 row accepted or a new Tier 2 row added (Group B, 4), new Tier 2 row written with full "How It Can Never Happen Again" prose at the rigor of Row 27a (Group C, 3).
- Adversarial tests in the same file prove the checker catches fabricated-gate and removed-cite regressions.
- `.test-floor-contract.json` bumped by the test-count delta in the same commit per the ratchet-bump rule.

### PR DESCRIPTION
**Pull request by oddessentials**
Closes #286.

## Summary

- **#286 contract-drift fix.** `scripts/run_pr_preflight.py` declares 38 unique `CommandSpec` names; 16 were missing as double-quoted strings from `LOCAL_CI_PARITY_INVARIANTS.md`, violating the doc's own claim to be the authoritative reference for named gates. New `tests/unit/test_parity_doc_coverage.py` AST-extracts preflight CommandSpecs and asserts the set is a subset of quoted strings in the doc — **verbatim name presence only**, no row/tier/prose enforcement. Adversarial negatives prove the checker catches both a fabricated new CommandSpec without a cite and a removed cite for an existing CommandSpec. Doc edits: Group A (9) and Group B (4) get minimal `CommandSpec "X"` citations on existing rows (7, 7a, 7b, 7c, 11, 11a, 11b, 11c, 26, 27, 32); Group C (3) gets new Tier 2 rows 35/36/37 for gates that had no row anywhere. `.test-floor-contract.json` Python floor bumped 2046 → 2052 (+6).

- **T025 dashboard-load overhead flake (`metrics.edge-cases.test.ts`).** Structural fix for the #202 recurrence observed on run 24681471103 (`Expected: < 10, Received: 131.83`). Root cause: small-base amplification — when `v1Ms` sits just above the noise floor, a few ms of jitter on `v2Ms` inflates `((v2-v1)/v1)*100` into meaningless three-digit values. Fix: introduce `NOISE_BAND_MS = 3 × NOISE_FLOOR_MS = 15ms` and `ABSOLUTE_CEILING_MS = 30ms`; fall back to the absolute-ms assertion whenever **either** measurement is in the band, only compute the percentage when both are well clear. Verified against the original 131% failure case (v1=5.1, v2=11.8 → now passes via absolute branch) plus edge cases that still fail real regressions (v2=35ms absolute breach, 94% percent breach).

- **parity-gate pip install retry (`ci.yml`).** Wrapped `pip install -e .[dev]` in a 4-attempt loop with exponential backoff (2s/4s/8s) to absorb transient PyPI TLS stream errors like the `ssl.SSLError: [SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC]` that tripped run 24677983042 mid-numpy-wheel download. pip's own `--retries` governs connection establishment only, not mid-stream decryption failures. Scoped to parity-gate; other pip-install sites haven't exhibited this failure and expanding scope would be churn.

## Test plan

- [x] `pytest tests/unit/test_parity_doc_coverage.py` — 6 passed (subset check + 2 adversarial negatives + 2 helper unit tests + baseline sanity)
- [x] `pnpm test tests/modules/metrics.edge-cases.test.ts` — 13/13 pass; telemetry confirms absolute-ms branch engages when either measurement is in the noise band
- [x] Broader `jest --testPathIgnorePatterns=python-integration ...` — 113 suites / 2463 tests pass
- [x] Synthetic edge-case sweep of the new T025 branching logic — 6/6 cases behave correctly (original flake passes, real regressions still fail)
- [x] `ci.yml` parses cleanly as YAML; no structural parity test asserts the shape of the parity-gate `Install dependencies` step
- [x] `.test-floor-contract.json` bumped by exactly the +6 test delta per the ratchet-bump rule

https://claude.ai/code/session_01A7ZXC3MMNVQzHiuSk43rbG

### CODE REVIEW

File: extension/tests/modules/metrics.edge-cases.test.ts [hunk 1]
```diff
@@ -635,11 +650,16 @@ describe("T025: SC-002 dashboard load time overhead", () => {
...
+    // Small-base amplification zone: if either measurement sits in the
+    // noise band, the percentage comparison is not a meaningful signal —
+    // a 5ms base with 3ms of jitter reads as 60% overhead with no real
+    // regression. Assert absolute-ms ceilings instead. The percentage
+    // path only runs when both measurements are well above the band, so
+    // its denominator is large enough that ordinary jitter cannot inflate
+    // the ratio past the 10% gate.
+    if (v1Ms < NOISE_BAND_MS || v2Ms < NOISE_BAND_MS) {
+      expect(v1Ms).toBeLessThan(ABSOLUTE_CEILING_MS);
+      expect(v2Ms).toBeLessThan(ABSOLUTE_CEILING_MS);
```
[the remark this answers is not in the document]
Comment (oddessentials) [reply extension/tests/modules/metrics.edge-cases.test.ts]: Good catch — acknowledged and fixed in f99ddf7. The previous rule keyed the fallback off per-measurement position (`v1 < NOISE_BAND_MS || v2 < NOISE_BAND_MS`), which, as you noted, let `v1=14, v2=29` (107% overhead, 15ms absolute delta) pass through the absolute ceiling.

Restructured to key the fallback off the absolute **delta** instead: `delta < NOISE_DELTA_MS (= 2 × NOISE_FLOOR_MS = 10ms)` → absolute-ms ceiling; otherwise → percentage branch. A 15ms delta is well outside the jitter envelope, so your case now correctly lands in the percentage branch and fails at 107%. The original #286 flake (v1=5.1, v2=11.8, delta=6.7) still passes via the absolute branch because 6.7ms of delta is in jitter territory regardless of how amplified the percentage reads.

Verified against 9 edge cases (including yours). The only residual "slips" are sub-10ms absolute deltas with high percentages — which I believe is the right call: at that scale the signal isn't distinguishable from GC/JIT noise on shared runners. The absolute ceiling (30ms) still catches runner-wide blowouts even in that branch.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/unit/test_parity_doc_coverage.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        lines = [
+            f'  - CommandSpec "{name}" — add a verbatim citation to the'
+            " relevant row in LOCAL_CI_PARITY_INVARIANTS.md"
+            for name in missing
+        ]
+        raise AssertionError(
+            "CommandSpec names missing from LOCAL_CI_PARITY_INVARIANTS.md:\n"
+            + "\n".join(lines)
+            + "\n\nEach preflight CommandSpec must appear verbatim as a "
+            "double-quoted string somewhere in the doc. This is a "
```
[the remark this answers is not in the document]
Comment (oddessentials) [reply tests/unit/test_parity_doc_coverage.py:88]: Declining intentionally. The terse error message is part of the test's design contract, not an oversight.

Per the #286 discussion, the whole value of the gate is that it enforces **verbatim name-presence only** and nothing about row location, tier, prose, or section heading. The moment the failure message starts telling contributors "add to the relevant Tier 1 or Tier 2 row," the test starts having an opinion about doc structure — which is exactly the scope creep the issue's design decisions call out as "doc-parity churn bait." If a contributor adds a new `CommandSpec` and the guard fires, the correct fix is "cite the name somewhere in the doc"; where to put it is a judgment call the doc author should make from the surrounding context, not one this test should prescribe.

The remediation template in the current message (`CommandSpec "<name>" — add a verbatim citation to the relevant row in LOCAL_CI_PARITY_INVARIANTS.md`) plus the explicit "This is a name-presence check only — row location, prose, and formatting are not enforced" footer is deliberately as much guidance as the test should offer.

---
_Generated by [Claude Code](https://claude.ai/code)_
